src/tls/test-connection.c

```diff
@@ -150,6 +157,33 @@ test_read_write (void)
 
   g_assert_cmpint (connection_read (c, WS), ==, RETRY);
 
+  /* test partial writes: send a large block */
+  memset (bigmsg, 42, sizeof (bigmsg));
+  g_assert_cmpint (send (client_fds[1], bigmsg, sizeof (bigmsg), 0), ==, sizeof (bigmsg));
+  g_assert_cmpint (connection_read (c, CLIENT), ==, SUCCESS);
+  g_assert_cmpint (c->buf_client.length, ==, sizeof (bigmsg));
+  /* write to ws should be partial due to our SO_SNDBUF from above */
+  g_assert_cmpint (connection_write (c, CLIENT), ==, PARTIAL);
+  g_assert_cmpint (c->buf_client.offset, >, 0);
+  g_assert_cmpint (c->buf_client.offset, <, sizeof (bigmsg));
+  /* flush the bigmsg to ws */
+  while (c->buf_client.offset > 0)
+    {
+      int len;
+
+      ConnectionResult r = connection_write (c, CLIENT);
+      if (r != SUCCESS && r != RETRY)
+        g_assert_cmpint (r, ==, PARTIAL);
+
+      for (;;)
+        {
+          len = recv (ws_fds[1], buffer, sizeof (buffer), 0);
+          if (len <= 0)
```

Comment:
"On error, terminate the test case, successfully".  Maybe handle the 0 and the -1 cases separately?

Reply:
Ack, done.